### ISSUE TITLE
[P2] 学習則ごとの重み更新・収束挙動差を視覚比較するデモページを実装する

### ISSUE THREAD

**Issue by tomiya7688** [tomiya7688/comfyUI_support_tools#98]
## 目的
数学的シミュレーション／AI学習デモの一部として、**同じデータ・同じ初期重み・同じモデルで、学習則だけを変えたときに重み更新や収束の仕方がどう変わるか**を視覚的に比較できるデモページを実装する。

#97 の活性化関数比較と対になる位置付けで、活性化関数側では「出力と勾配の形」、本Issueでは**「その誤差や勾配をどう使って重みを更新するか」**に焦点を当てる。

関連:
- #92 機械学習の学習ループ
- #94 微分・積分・偏微分・連鎖律
- #97 活性化関数ごとの学習挙動差比較

---

## 比較対象
MVPでは、教育的に違いが分かりやすい以下を扱う。

- Perceptron learning rule
- Delta rule / LMS
- Gradient Descent
- Stochastic Gradient Descent (SGD)

追加候補:
- Momentum
- Adam

Hebbian learningは教師あり誤差学習と性質が異なるため、MVPでは補足または別モードとして扱う。

---

### 1. 1サンプルでの更新則比較
同じ入力・正解・予測・初期重みから、学習則ごとの更新式を並べて表示する。

例:
```text
input: x
prediction: y_hat
target: y
error: e = y - y_hat
```

各ruleで以下を表示:
- 現在の誤差
- 更新式
- 各weightのdelta
- biasのdelta
- 更新前 / 更新後の値

例:
```text
w_new = w_old + Δw
```

- [ ] 1 Stepボタンで各ruleを同時更新
- [ ] 変更された重みを強調表示
- [ ] 更新量を矢印やbarで比較

---

### 2. 決定境界の動き比較
2D分類データを使い、各学習則でdecision boundaryがどう動くかを表示する。

- [ ] 初期境界を共通化
- [ ] 1 stepごとの境界移動を表示
- [ ] epochごとの境界を履歴表示可能にする
- [ ] 学習則ごとに並列表示または切替表示

視覚的には、
```text
same initial boundary
  ↓
learning rule A -> move this way
learning rule B -> move differently
```
を確認できるようにする。

---

### 3. 複数epoch比較
同じdataset / seed / initial weights / learning rate条件で各学習則を実行する。

表示:
- loss curve
- accuracy curve
- weight trajectory
- bias trajectory
- update magnitude
- decision boundary

- [ ] Run / Pause / Reset
- [ ] epochをstep送りできる
- [ ] 同じ初期状態へ完全Resetできる

---

## 学習率の操作
学習則そのものの違いに加え、`learning rate` の役割も簡単に見せる。

- [ ] learning rate slider
- [ ] 小さすぎる場合の遅い更新
- [ ] 大きすぎる場合の振動 / 発散を視覚化
- [ ] 同じruleでlearning rateだけ変えた比較も可能にする

ただし本Issueの主題はlearning rate比較ではなく、あくまで学習則比較とする。

---

## Perceptron rule
二値分類で、誤分類時に重みを更新する仕組みを表示する。

例:
```text
Δw = η (y - y_hat) x
```

説明:
- 「誤分類したとき、正しい側へ境界を押す」
- 正しく分類できた場合に更新しないケースがあることを示す

---

## Delta rule / LMS
出力誤差の大きさを使って連続的に重みを修正する。

例:
```text
E = 1/2 (y - y_hat)^2
Δw ∝ -∂E/∂w
```

説明:
- 「間違ったかどうかだけでなく、どれくらい外れたかを使って更新する」

#94 の微分・偏微分につなげる。

---

## Gradient Descent
loss surface上で現在のparameter位置とgradient directionを簡易表示する。

- [ ] 2 parameter程度のloss surface / contourを表示
- [ ] 現在位置をpointで表示
- [ ] gradient vectorを表示
- [ ] 1 stepで反対方向へ動くことを表示

```text
θ_new = θ_old - η ∇L(θ)
```

説明:
- 「損失が最も増える方向が勾配」
- 「学習ではその反対へ進む」

---

## SGD
Gradient Descentとの差をデータ選択単位で見せる。

- Batch Gradient Descent:
  - 全データから1回の更新方向を計算
- SGD:
  - 1サンプルまたは小さいbatchごとに更新

視覚化:
- Batch GD: なめらかなtrajectory
- SGD: 少し揺れながら進むtrajectory

- [ ] 現在使っているsampleを強調表示
- [ ] sampleが変わるとgradient方向が変化することを確認可能にする

---

## Momentum / Adam（追加候補）
余裕があれば追加する。

### Momentum
過去の更新方向を持つことで、谷方向へ進みやすくなる様子を表示。

### Adam
1次モーメント / 2次モーメントの概念を簡略表示し、parameterごとに実効step幅が変化することを示す。

MVPでは内部式の完全な厳密解説より、更新挙動の違いを見せることを優先する。

---

## 学習データ
理解しやすさを優先し、2Dの小規模データを利用する。

候補:
- linearly separable
- noisy linearly separable
- XOR風データ（線形モデルでは解けない例として補足）

XORについては「学習則を変えてもモデル表現力が足りなければ解けない」ことを示せるとよい。

---

## 共通UI
- [ ] 学習則selector
- [ ] 複数rule同時比較
- [ ] learning rate slider
- [ ] 1 Step
- [ ] Run
- [ ] Pause
- [ ] Reset
- [ ] epoch表示
- [ ] current sample表示（SGD時）
- [ ] weight / bias表示
- [ ] loss / accuracy表示
- [ ] decision boundary表示

---

## 教育用説明
短い説明を表示する。

例:
- 「学習則は、誤差を見たあと重みをどう変えるかを決めます」
- 「同じ誤差でも、更新方法が違えば学習経路も変わります」
- 「SGDは一部のデータで少しずつ更新するため、経路が揺れることがあります」
- 「勾配降下法は損失が小さくなる方向へパラメータを動かします」
- 「学習則を変えても、モデル自体が表現できない問題は解けません」

---

## #97との接続
#97と役割を明確に分ける。

```text
Activation function
  -> output / local derivative

Learning rule
  -> parameter update policy
```

ページ間リンクを用意し、

```text
forward calculation
  ↓
loss
  ↓
gradient
  ↓
learning rule
  ↓
weight update
```

という流れを追えるようにする。

---

## テスト
- [ ] Perceptron ruleで誤分類時に期待したweight更新になる
- [ ] Delta ruleで誤差量に応じた更新になる
- [ ] Gradient Descentがnegative gradient方向へparameter更新する
- [ ] SGDが選択sampleに応じたgradientで更新する
- [ ] 同一seed / initial weights / datasetで比較条件が揃う
- [ ] learning rate変更で更新量が変化する
- [ ] Resetで全ruleが同じ初期状態へ戻る
- [ ] epoch / loss / boundary表示が更新に追従する

---

## 完了条件
- 同じ初期条件で学習則だけ変えて比較できる
- 1 stepごとのweight更新式と数値を確認できる
- decision boundaryの動きの違いを確認できる
- Batch GDとSGDのtrajectory差を視覚的に確認できる
- 学習率が更新量へ与える影響を操作できる
- 「学習則 = 誤差/勾配を使ってパラメータをどう更新するか」という役割がページ単体で伝わる
- #97の活性化関数デモと組み合わせて、学習全体の仕組みを追える

## Out of Scope
- 大規模ニューラルネットワーク
- 分散学習
- optimizer性能benchmark
- AutoML / optimizer自動選択
- GPU必須化
- 学習則の優劣ランキング
- 高度な2次最適化法
